offchain/jd/cognito.go

```
@@ -0,0 +1,144 @@
+// This interface allows for mocking the AWS Cognito client in unit tests.
+type CognitoClient interface {
+	InitiateAuth(
+		ctx context.Context,
+		params *cognitoidentityprovider.InitiateAuthInput,
+		optFns ...func(*cognitoidentityprovider.Options),
+	) (*cognitoidentityprovider.InitiateAuthOutput, error)
+}
+
+// CognitoTokenSource provides a oath2 token that is used to authenticate with the Job Distributor.
+type CognitoTokenSource struct {
+	// The Cognito authentication information
+	auth CognitoAuth
+
+	// The cached authentication result from Cognito
+	authResult *types.AuthenticationResultType
+
+	// The Cognito client interface for making API calls
+	client CognitoClient
+}
+
+// CognitoAuth contains the Cognito authentication information required to generate a token from
+// Cognito.
+type CognitoAuth struct {
+	AWSRegion              string
+	CognitoAppClientID     string
+	CognitoAppClientSecret string
+	Username               string
+	Password               string
+}
+
+// NewCognitoTokenSource creates a new CognitoTokenSource with the given CognitoAuth configuration.
+// If client is nil, a real AWS Cognito client will be created when Authenticate is called.
+func NewCognitoTokenSource(auth CognitoAuth) (*CognitoTokenSource, error) {
+	return &CognitoTokenSource{
+		auth: auth,
+	}, nil
+}
+
+// Authenticate performs user authentication against AWS Cognito using the USER_PASSWORD_AUTH flow.
+//
+// Authentication results are cached in the authResult field and used by the Token() method
+// to provide OAuth2 access tokens without re-authenticating.
+func (c *CognitoTokenSource) Authenticate(ctx context.Context) error {
+	// Create client if not already set (for production use)
+	if c.client == nil {
+		sdkConfig, err := config.LoadDefaultConfig(ctx,
+			config.WithRegion(c.auth.AWSRegion),
+		)
+		if err != nil {
+			return err
+		}
+		c.client = cognitoidentityprovider.NewFromConfig(sdkConfig)
```

i wonder should this be done on the `NewCognitoTokenSource` stage so we know early if there is issue with the config? `NewCognitoTokenSource` also supports returning an error whihc we are not using now.

---

The original implementation had this one the constructor, but I moved all the setup out into an Authenticate function.

The user has two options here:

1. Very simple with little control 

```
s := NewCognitoTokenSource(auth)
s.Token()
```

This will handle everything for them under the hood, but they do not get to fail early when requesting the token.

2. Fine grained control

```
s := NewCognitoTokenSource(auth)
err := s.Authenticate(ctx) // Authenticate early to capture any errors

// .... do whatever they want

s.Token()
```

Coupling it into the constructor means that you must now always pass a context to the constructor which I was trying to avoid.